### PR TITLE
[new resources and datasources for cynosdb]

### PR DESCRIPTION
**Pull request by brickzzhang**
resource cluster:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbClusterResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbClusterResource -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	0.466s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbClusterResource
--- PASS: TestAccTencentCloudCynosdbClusterResource (173.80s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	174.463s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

-------

resource readonly instance:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbReadonlyInstanceResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbReadonlyInstanceResource -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	1.072s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbReadonlyInstanceResource
--- PASS: TestAccTencentCloudCynosdbReadonlyInstanceResource (155.74s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	156.584s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

------

datasource cluster:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbClustersDataSource_full"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbClustersDataSource_full -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	0.964s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbClustersDataSource_full
=== PAUSE TestAccTencentCloudCynosdbClustersDataSource_full
=== CONT  TestAccTencentCloudCynosdbClustersDataSource_full
--- PASS: TestAccTencentCloudCynosdbClustersDataSource_full (64.62s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	65.477s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

------

datasource instance:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbInstancesDataSource_full"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbInstancesDataSource_full -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	(cached) [no tests to run]
=== RUN   TestAccTencentCloudCynosdbInstancesDataSource_full
=== PAUSE TestAccTencentCloudCynosdbInstancesDataSource_full
=== CONT  TestAccTencentCloudCynosdbInstancesDataSource_full
--- PASS: TestAccTencentCloudCynosdbInstancesDataSource_full (122.43s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	123.229s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

### CODE REVIEW

File: examples/tencentcloud-cynosdb/main.tf [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    "Sat",
+    "Sun",
+    "Thu",
+    "Wed",
+    "Tue",
+  ]
+}
+
+data "tencentcloud_cynosdb_clusters" "foo" {
+  cluster_id   = "cynosdbmysql-dzj5l8gz"
```
Comment (gailwang) [inline examples/tencentcloud-cynosdb/main.tf:64]: 使用resource.id

File: examples/tencentcloud-cynosdb/main.tf [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  project_id   = 0
+  db_type      = "MYSQL"
+  cluster_name = "test"
+}
+
+data "tencentcloud_cynosdb_instances" "foo" {
+  instance_id   = "cynosdbmysql-ins-0wln9u6w"
```
Comment (gailwang) [inline examples/tencentcloud-cynosdb/main.tf:71]: 同上，我写的我认领

File: tencentcloud/extension_cynosdb.go [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+		"storage_used": {
+			Type:        schema.TypeInt,
+			Computed:    true,
+			Description: "Used storage of CynosDB cluster, unit in MB.",
+		},
+		// rw instance group infos
+		"rw_group_id": {
+			Type:        schema.TypeString,
+			Computed:    true,
+			Description: "Id of read-write instance group.",
```
Comment (gailwang) [inline tencentcloud/extension_cynosdb.go:213]: ID id Id统一

File: tencentcloud/extension_cynosdb.go [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+		"prepaid_period": {
+			Type:         schema.TypeInt,
+			Optional:     true,
+			ValidateFunc: validateAllowedIntValue(CYNOSDB_PREPAID_PERIOD),
+			Description:  "The tenancy (time unit is month) of the prepaid instance. Valid values are 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 24, 36. NOTE: it only works when charge_type is set to `PREPAID`.",
+		},
+		"auto_renew_flag": {
+			Type:        schema.TypeInt,
+			Optional:    true,
+			Description: "Auto renew flag. Valid values are `0`(MANUAL_RENEW), `1`(AUTO_RENEW). Only works for PREPAID cluster.",
```
Comment (gailwang) [inline tencentcloud/extension_cynosdb.go:175]: 这个参数有default吗

File: tencentcloud/resource_tc_cynosdb_cluster_test.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+					resource.TestCheckResourceAttr("tencentcloud_cynosdb_cluster.foo", "tags.test", "test-update"),
+					resource.TestCheckResourceAttr("tencentcloud_cynosdb_cluster.foo", "rw_group_sg.#", "1"),
+					resource.TestCheckResourceAttr("tencentcloud_cynosdb_cluster.foo", "ro_group_sg.#", "1"),
+				),
+			},
+			{
+				ResourceName:            "tencentcloud_cynosdb_cluster.foo",
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"password", "auto_renew_flag", "force_delete", "storage_limit"},
```
Comment (gailwang) [inline tencentcloud/resource_tc_cynosdb_cluster_test.go:75]: auto_renew_flag为啥import不进来？

File: tencentcloud/resource_tc_cynosdb_readonly_instance.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+	)
+
+	// check unsupported field modification
+	unsupported := []string{
+		"cluster_id",
+		"instance_name",
+	}
+	for _, v := range unsupported {
+		if d.HasChange(v) {
+			return fmt.Errorf("[CRITAL] field %s is not allowed to be modified", v)
```
Comment (gailwang) [inline tencentcloud/resource_tc_cynosdb_readonly_instance.go:206]: 这些字段如果不能修改 特别是cluster_id为啥不写成forceNew

**Comment by brickzzhang** [comment]
cluster:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbClusterResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbClusterResource -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	(cached) [no tests to run]
=== RUN   TestAccTencentCloudCynosdbClusterResource
--- PASS: TestAccTencentCloudCynosdbClusterResource (253.35s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	254.838s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]


------------

instance:

➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbReadonlyInstanceResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbReadonlyInstanceResource -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	0.987s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbReadonlyInstanceResource
--- PASS: TestAccTencentCloudCynosdbReadonlyInstanceResource (158.97s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	159.807s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]


-------------

single apply:

tencentcloud_security_group.sglab: Creating...
tencentcloud_vpc.foo: Creating...
tencentcloud_vpc.foo: Creation complete after 2s [id=vpc-eztqle3k]
tencentcloud_subnet.subnet: Creating...
tencentcloud_security_group.sglab: Creation complete after 2s [id=sg-8hsohxqm]
tencentcloud_subnet.subnet: Creation complete after 1s [id=subnet-atauea6z]
tencentcloud_cynosdb_cluster.foo: Creating...
tencentcloud_cynosdb_cluster.foo: Still creating... [10s elapsed]
tencentcloud_cynosdb_cluster.foo: Still creating... [20s elapsed]
tencentcloud_cynosdb_cluster.foo: Still creating... [30s elapsed]
tencentcloud_cynosdb_cluster.foo: Still creating... [40s elapsed]
tencentcloud_cynosdb_cluster.foo: Still creating... [50s elapsed]
tencentcloud_cynosdb_cluster.foo: Creation complete after 55s [id=cynosdbmysql-8jfwr50q]
tencentcloud_cynosdb_readonly_instance.foo: Creating...
tencentcloud_cynosdb_readonly_instance.foo: Still creating... [10s elapsed]
tencentcloud_cynosdb_readonly_instance.foo: Still creating... [20s elapsed]
tencentcloud_cynosdb_readonly_instance.foo: Still creating... [30s elapsed]
tencentcloud_cynosdb_readonly_instance.foo: Still creating... [40s elapsed]
tencentcloud_cynosdb_readonly_instance.foo: Creation complete after 50s [id=cynosdbmysql-ins-4tzcowbd]

Apply complete! Resources: 5 added, 0 changed, 0 destroyed.


------------

multiple apply:

➜  200904-cynosdb tf apply
2020/09/22 17:10:22 [WARN] Log levels other than TRACE are currently unreliable, and are supported only for backward compatibility.
  Use TF_LOG=TRACE to see Terraform's internal logs.
  ----
tencentcloud_security_group.sglab: Refreshing state... [id=sg-8hsohxqm]
tencentcloud_vpc.foo: Refreshing state... [id=vpc-eztqle3k]
tencentcloud_subnet.subnet: Refreshing state... [id=subnet-atauea6z]
tencentcloud_cynosdb_cluster.foo: Refreshing state... [id=cynosdbmysql-8jfwr50q]
tencentcloud_cynosdb_readonly_instance.foo: Refreshing state... [id=cynosdbmysql-ins-4tzcowbd]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

**Review by gailwang — APPROVED** [review]

**Comment by brickzzhang** [comment]
image

File: tencentcloud/data_source_tc_cynosdb_clusters.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	clusterType := ""
+	if v, ok := d.GetOk("cluster_type"); ok {
+		clusterType = v.(string)
+	}
+
+	cynosdbService := CynosdbService{
+		client: meta.(*TencentCloudClient).apiV3Conn,
+	}
+
+	err := resource.Retry(readRetryTimeout, func() *resource.RetryError {
```
Comment (oliverpei) [inline tencentcloud/data_source_tc_cynosdb_clusters.go:166]: retry最好只包含api请求

File: tencentcloud/resource_tc_cynosdb_cluster.go [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+	// check unsupported field modification
+	unsupported := []string{
+		"project_id", "vpc_id", "subnet_id", "port", "storage_limit", "cluster_name",
+		"password", "prepaid_period", "auto_renew_flag",
+	}
+	for _, v := range unsupported {
+		if d.HasChange(v) {
+			return fmt.Errorf("[CRITAL] field %s is not allowed to be modified", v)
+		}
+	}
```
Comment (oliverpei) [inline tencentcloud/resource_tc_cynosdb_cluster.go:409]: 这种尽量少写，不能修改就ForceNew

**Comment by brickzzhang** [comment]
➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbClustersDataSource_full"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbClustersDataSource_full -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	3.625s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbClustersDataSource_full
=== PAUSE TestAccTencentCloudCynosdbClustersDataSource_full
=== CONT  TestAccTencentCloudCynosdbClustersDataSource_full
--- PASS: TestAccTencentCloudCynosdbClustersDataSource_full (90.59s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	93.831s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]


----------


➜  terraform-provider-tencentcloud git:(master) ✗ make testacc TESTARGS="-run=TestAccTencentCloudCynosdbClusterResource"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccTencentCloudCynosdbClusterResource -timeout 120m
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud	[no test files]
testing: warning: no tests to run
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/gendoc	0.259s [no tests to run]
=== RUN   TestAccTencentCloudCynosdbClusterResource
--- PASS: TestAccTencentCloudCynosdbClusterResource (264.29s)
PASS
ok  	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud	264.797s
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/connectivity	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/internal/helper	[no test files]
?   	github.com/tencentcloudstack/terraform-provider-tencentcloud/tencentcloud/ratelimit	[no test files]

**Review by oliverpei — APPROVED** [review]

**Pull request merged**